### PR TITLE
✨😈☝️ pull request preview playground

### PR DESCRIPTION
**Pull request by carstingaxion**
### Description of the Change

grafik

Closes #

### Changelog Entry

> Added - New automated comments with links to preview the current PR in WordPress Playground.

### Credits

Props @carstingaxion 

### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/GatherPress/gatherpress/blob/main/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

### CODE REVIEW

**Comment by carstingaxion** [comment]
grafik


>Artifact gatherpress-plugin.zip successfully finalized. Artifact ID 1557248140
Artifact gatherpress-plugin has been successfully uploaded! Final size is 19794406 bytes. Artifact ID is 1557248140
Artifact download URL: https://github.com/GatherPress/gatherpress/actions/runs/9322864932/artifacts/1557248140

**Comment by carstingaxion** [comment]
grafik

**Comment by carstingaxion** [comment]
>      message: 'Resource not accessible by integration',
>      documentation_url: 'https://docs.github.com/rest/issues/comments#create-an-issue-comment'

**Comment by carstingaxion** [comment]
> [What and Why pull_request_target?](https://dev.to/suzukishunsuke/secure-github-actions-by-pullrequesttarget-641)
>
>[pull_request_target](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) is one of the events triggering GitHub Actions workflows.
One of the differences between pull_request_target and pull_request is that pull_request_target triggers workflows based on the latest commit of the pull request's base branch.
Even if workflow files are modified or deleted on feature branches, workflows on the default branch aren't affected so you can prevent malicious code from being executed in CI without code review.
>
> [...]
>
> The GitHub Actions [built in environment variables](https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables) and [Context](https://docs.github.com/en/actions/learn-github-actions/contexts) of pull_request_target event are different from those of pull_request event.
For example, the following environment variables and context are different.
>
>    event_name, GITHUB_EVENT_NAME
    ref, GITHUB_REF
    sha, GITHUB_SHA
    ref_name, GITHUB_REF_NAME

**Comment by carstingaxion** [comment]
grafik

**Comment by carstingaxion** [comment]
```mermaid
sequenceDiagram
   participant PR as pull_request
 
    loop 'opened', 'updated'
    create participant zip as zip artifact workflow
    PR-->>+zip: Trigger action

    zip-->>zip: Get latest commit from PR
    zip-->>zip: Generate .zip of current state
    zip->>-PR: Save zip as artifact to workflow run
    zip-->>+zip: Finish workflow run

    create participant blueprint as blueprint workflow
    zip-->>+blueprint: Triggers action
    destroy zip
    zip-->>+PR: Wait for next change
    end
    destroy blueprint
    blueprint-->>PR: Not a Comment,but a console.log()<br/><br/>with a link to playground 
    create actor user as @user
    PR-->>user: 👀 reads link from console.log() on the workflow run
    create participant playground
    user->>playground: start

```

**Comment by carstingaxion** [comment]
grafik

**Comment by carstingaxion** [comment]
_**No way**_, [this already exists](https://github.com/WordPress/wordpress-playground/issues/1237) and is even [semi-documented](https://github.com/WordPress/gutenberg/issues/60795) (at least on Github).

I'm fiddling two days now, to do exact this _on my own_. Ohhhh. 🤦

**Comment by carstingaxion** [comment]
| v1 | v2 |
|--------|--------|
| own infrastracture on gatherpress.org | depends on github-proxy.com |
| 1.3 MB incl. BUILD | 40 MB, No BUILD |
| much new code to maintain | minimal new code |
| blueprint is a little longer because artifact-unzip-ing | blueprint is simple |
| [RESULT](https://github.com/GatherPress/gatherpress/actions/runs/9346372869/job/25721126162) *(open the **Run...** step)* | [RESULT](https://github.com/carstingaxion/gatherpress/pull/23#issuecomment-2144781778) |

**Comment by carstingaxion** [comment]
>40 MB, No BUILD

Just realised that, if we delete the `.mp4` from `/docs/media`, like Patricia suggested, we can safe a lot.

File: .gatherpress-org/playground-preview/plugin-proxy.php
Comment (carstingaxion) [inline .gatherpress-org/playground-preview/plugin-proxy.php:1]: Needs to be uploaded to `gatherpress.org`

```
└── playground-preview
    ├── .htaccess
    └── plugin-proxy.php
```

File: .gatherpress-org/playground-preview/.htaccess
Comment (carstingaxion) [inline .gatherpress-org/playground-preview/.htaccess:1]: Needs to be uploaded to `gatherpress.org`

```
└── playground-preview
    ├── .htaccess
    └── plugin-proxy.php
```

File: .gatherpress-org/playground-preview/plugin-proxy.php [hunk 1]
```diff
@@ -0,0 +1,322 @@
+<?php
+/**
+ * https://gatherpress.org/some-random-chars/plugin-proxy.php?org=GatherPress&repo=gatherpress&workflow=Build%20GatherPress%20Plugin%20Zip&artifact=gatherpress-pr&pr=666
+ * 
+ * 
+ * Based on: https://github.com/WordPress/wordpress-playground/blob/ca759ee3281837596bfdd9736bbea205104741b2/packages/playground/website/public/plugin-proxy.php
+ */
+ini_set( 'display_errors', 0 );
+
+putenv( 'GITHUB_TOKEN=___REPLACE_WITH_REAL_GITHUB_TOKEN___' );
```
Comment (carstingaxion) [inline .gatherpress-org/playground-preview/plugin-proxy.php:10]: Replace `___REPLACE_WITH_REAL_GITHUB_TOKEN___` with a working PAT (to comment on PRs).

Only in PRODUCTION on gatherpress.org, NOT here in the repo!

File: .github/scripts/create-preview-links.js [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+ * @param {object} context - The context of the event that triggered the action.
+ * @param {number} number - The PR number where the plugin changes are located.
+ * @returns {string} - The URL where to download the built & zipped plugin, as an artifact of the workflow.
+ */
+function createBlueprintUrl(context, number) {
+	const { repo, owner } = context;
+	const workflow = encodeURI('Playground Preview');
+	const artifact = 'gatherpress-pr';
+	const proxy = 'https://hub.carsten-bach.de/gatherpress/plugin-proxy.php';
+	// const proxy = 'https://gatherpress.org/playground-preview/plugin-proxy.php';
```
Comment (carstingaxion) [inline .github/scripts/create-preview-links.js:18]: Swap out `hub.carsten-bach.de...` against `gatherpress.org...`, 
as soon as `gatherpress.org/playground-preview/plugin-proxy.php` is running.

**Comment by carstingaxion** [comment]
| _Preview-Comment_ after last commit | A new commit is added | The _Preview-Comment_ is deleted & recreated after the last commit |
|--------|--------|--------|
| grafik | grafik | grafik | 

As the screenshots show: **The playground PR previewer is ready and working!**

Actually this is not yet running on the `GatherPress/gatherpress` repository, because the workflow needs to be part of the repo itself _BEFORE_ it will run & work. To circumvent this behaviour during development I created https://github.com/carstingaxion/gatherpress/pull/21, which is the working example and origin of the screenshots.

File: .gatherpress-org/playground-preview/plugin-proxy.php
Comment (mauteri) [reply .gatherpress-org/playground-preview/plugin-proxy.php:1]: @MervinHernandez you'll need to work with Carstin on this to get on gatherpress.org

**Review by mauteri — APPROVED** [review]

File: .github/scripts/create-preview-links.js (hunk 2 above)
Comment (carstingaxion) [reply .github/scripts/create-preview-links.js:18]: Is part of https://github.com/GatherPress/gatherpress/pull/741

**Pull request merged**